src/ynEIGEN/withdrawalsProcessor.sol

```diff
@@ -0,0 +1,251 @@
+
+    struct QueuedWithdrawal {
+        address node;
+        address strategy;
+        uint256 nonce;
+        uint256 shares;
+        uint32 startBlock;
+        bool completed;
+    }
+
+    // @todo - put ids in a struct
+    uint256 public queuedId;
+    uint256 public completedId;
+    uint256 public processedId;
+
+    uint256 public totalQueuedWithdrawals;
+
+    uint256 public minPendingWithdrawalRequestAmount;
+
+    // yieldnest
+    IWithdrawalQueueManager public immutable withdrawalQueueManager;
+    ITokenStakingNodesManager public immutable tokenStakingNodesManager;
+    IAssetRegistry public immutable assetRegistry;
+    IYieldNestStrategyManager public immutable ynStrategyManager;
+
+    // eigenlayer
+    IDelegationManager public immutable delegationManager;
+
+    mapping(uint256 id => QueuedWithdrawal) public queuedWithdrawals;
+    mapping(uint256 fromId => uint256 toId) public batch;
+    mapping (IERC20 asset => uint256 minShares) public minShares;
+
+    //
+    // Constructor
+    //
+
+    constructor(
+        address _owner,
+        address _withdrawalQueueManager,
+        address _tokenStakingNodesManager,
+        address _assetRegistry,
+        address _ynStrategyManager,
+        address _delegationManager
+    ) Ownable(_owner) {
+        if (
+            _withdrawalQueueManager == address(0) ||
+            _tokenStakingNodesManager == address(0) ||
+            _assetRegistry == address(0) ||
+            _ynStrategyManager == address(0) ||
+            _delegationManager == address(0)
+        ) revert InvalidInput();
+
+        withdrawalQueueManager = IWithdrawalQueueManager(_withdrawalQueueManager);
+        tokenStakingNodesManager = ITokenStakingNodesManager(_tokenStakingNodesManager);
+        assetRegistry = IAssetRegistry(_assetRegistry);
+        ynStrategyManager = IYieldNestStrategyManager(_ynStrategyManager);
+        delegationManager = IDelegationManager(_delegationManager);
+
+        minPendingWithdrawalRequestAmount = 0.1 ether;
+
+        // @todo - fix that
+        IERC20[] memory _assets = assetRegistry.getAssets();
+        for (uint256 i = 0; i < _assets.length; ++i) {
+            minShares[_assets[i]] = 1 ether;
+        }
+    }
+
+    //
+    // Processor functions
+    //
+
+    /// @notice Queues withdrawals
+    /// @dev Reverts if the total pending withdrawal requests are below the minimum threshold
+    /// @dev Skips nodes with shares below the minimum threshold
+    /// @dev If a nodes has more shares than the minimum threshold, exits the node entirely
+    /// @dev Tries to satisfy the pending withdrawal requests while prioritizing withdrawals in one asset
+    /// @dev Saves the queued withdrawals together in a batch, to be completed in the next step (`completeQueuedWithdrawals`)
+    /// @return True if all pending withdrawal requests were queued, false otherwise
+    function queueWithdrawals() external onlyOwner returns (bool) {
```

Comment:
One requirement here is we keep nodes balanced

We can't empty one and leave the other ones filled up

We currently delegate stake to multiple operators, and this would imbalance that delegation

We can't fully empty a node as this currently does things.

I think a logic that gets the "full picture" of asset distribution and then takes assets from each node is wha we want

we can consider withdrawing the dominant asset (right now it's oeth) to save some gas (maybe some logic that tries to keep an asset ratio?)

Reply:
To save gas:

we can round robin through them to not explode gas cost in this call.

So we can withdraw potentially in 32 ETH (or 10 ETH) units in the same way as the native ETH.

so for example if you get a 5 ETH withdrawal  queued you just get it from node 0

but next time you try again, you will get it from node 1, because node 0 has less than the rest 4 nodes.